### PR TITLE
Add QEMU microvm hypervisor backend

### LINKED ISSUE #275: QEMU microvm support
Is there any chance to add support for QEMU microvm machine type?

Microvm machine type would permit faster boots and somewhat lower overhead.

If allowed, I would be happy to open a draft PR for review and testing.

### PR DESCRIPTION
**Pull request by rgarcia**
## Summary

- add `qemu-microvm` as a first-class hypervisor backend on Linux amd64
- use QEMU's `microvm` board with virtio-mmio disks, networking, vsock, and balloon devices
- keep standard `qemu` behavior unchanged (`q35` on amd64, `virt` on arm64)
- validate microvm constraints before side effects: no PCI/vGPU, no hotplug memory, and at most eight virtio-mmio devices
- preserve the backend across stop/start, standby/restore, snapshots, and warm forks
- add an opt-in q35-vs-microvm boot/RSS benchmark and user/developer documentation

The public API exposes only `hypervisor: qemu-microvm`; QEMU machine types remain an internal restore detail. This avoids adding a QEMU-specific `machine_type` field to the generic instance API.

## Background

QEMU's [`microvm` machine type](https://www.qemu.org/docs/master/system/i386/microvm.html) is a minimalist x86 virtual platform inspired by Firecracker. Compared with the general-purpose `q35` PC model, it deliberately omits PCI and ACPI, exposes paravirtualized devices over `virtio-mmio`, and supports at most eight of those devices. QEMU designed it for short-lived guests and as a baseline for optimizing boot time and footprint.

That model fits Hypeman's direct-kernel Linux guests: we already control the kernel/initrd and primarily need virtio disks, networking, vsock, and ballooning. Selecting `qemu-microvm` avoids initializing hardware that these workloads do not use while retaining Hypeman's existing QEMU/QMP lifecycle, snapshot, and fork machinery.

It is not a transparent replacement for standard QEMU. Upstream documents no PCI-only devices, no hotplug, and no live migration across QEMU versions. This PR therefore exposes it as a distinct Hypeman hypervisor capability contract instead of a generic QEMU board toggle, and validates those constraints before creating a VM.

Further reading:

- [QEMU: `microvm` virtual platform](https://www.qemu.org/docs/master/system/i386/microvm.html) — authoritative device model, options, and limitations
- [QEMU: x86 PC system models](https://www.qemu.org/docs/master/system/i386/pc.html) — background on the standard `pc`/`q35` machine families
- [Firecracker design](https://github.com/firecracker-microvm/firecracker/blob/main/docs/design.md) and the [NSDI 2020 Firecracker paper](https://www.usenix.org/conference/nsdi20/presentation/agache) — useful context for the microVM model that inspired QEMU's implementation

## Validation

- `go build ./...`
- `go vet ./lib/hypervisor/... ./lib/instances/... ./cmd/api/api/...`
- `go test ./lib/hypervisor/qemu ./cmd/api/api -count=1`
- focused instance unit tests for backend resolution and snapshot normalization
- `make test TEST='TestQEMUMicroVM.*' TEST_TIMEOUT=12m`
- standard QEMU regressions:
  - `TestQEMUBasicEndToEnd`
  - `TestQEMUStandbyAndRestore`
  - `TestQEMUForkFromRunningNetwork`
- real API/CLI smoke: `qemu-microvm` booted Alpine, reached `Running`, and returned `x86_64`; standard `qemu` continued to use q35
- OpenAPI regeneration is deterministic and no public `machine_type` remains

## Performance sample

Three sequential nginx samples on this shared Linux amd64/KVM host:

| backend | boot-to-program p50 | p95 | RSS p50 | RSS p95 |
| --- | ---: | ---: | ---: | ---: |
| qemu | 946ms | 1.00s | 324MB | 360MB |
| qemu-microvm | 381ms | 428ms | 333MB | 335MB |

A separate `./scripts/benchmark-microvm-hypervisors.sh 1` smoke run validated the QEMU-microvm vs Firecracker comparison path:

| backend | boot-to-program | VMM RSS |
| --- | ---: | ---: |
| qemu-microvm | 462.9ms | 330.3MB |
| firecracker | 439.9ms | 200.0MB |

The cross-hypervisor result is `n=1` and is included only as script validation, not as a performance conclusion. All benchmarks are opt-in and non-gating.

Closes #275

---

> [!NOTE]
> **Medium Risk**
> Touches core VM lifecycle, QEMU snapshots/restores, and version-locked memory restore for both QEMU backends; mis-validation or pool/version edge cases could break creates or restores, but scope is bounded by new backend plus shared QEMU paths with added tests.
> 
> **Overview**
> Introduces **`qemu-microvm`** as a first-class hypervisor (Linux **amd64** only): QEMU's **`microvm`** board with **virtio-mmio** devices instead of PCI. Standard **`qemu`** stays on **q35/virt**. Users select the backend via API/CLI (`hypervisor: qemu-microvm`); QEMU machine type is internal (persisted in **`qemu-config.json`** with **`QEMUVersion`**).
> 
> The QEMU implementation is split into **Standard** and **MicroVM profiles** sharing QMP/process code. **MicroVM** rejects PCI/vGPU, hotplug memory, and more than **eight** virtio-mmio slots. **`VMStarter`** gains **`ValidateConfig`** and **`ResolveVersion`**; instance create and stopped-snapshot restore/fork run **preflight** against the full disk/network/PCI/balloon plan before mutating state. **`RequiresHostSnapshotVersion`** ties QEMU standby/warm restore to the **exact host QEMU** that wrote the memory image; cold start and **stopped** snapshots can still switch between **`qemu`** and **`qemu-microvm`** when the plan validates.
> 
> Docs/Makefile cover **qemu-system** deps, backend limits, and opt-in boot benchmarks; integration tests cover microvm E2E, standby/restore, forks, and API mapping.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit b4a04e3747c26f2816c9522df3846f89a107db75. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lib/hypervisor/qemu/pool.go [hunk 1]
```diff
@@ -31,11 +42,14 @@ func GetOrCreate(socketPath string) (*QEMU, error) {
 
 	// Double-check after acquiring write lock
 	if client, ok := clientPool.clients[socketPath]; ok {
+		if client.hypervisorType != hypervisorType {
+			return nil, poolTypeMismatchError(socketPath, client.hypervisorType, hypervisorType)
+		}
```
[the remark this answers is not in the document]
Comment (rgarcia) [reply lib/hypervisor/qemu/pool.go:48]: Fixed in 33cffb15. QEMU startup now synchronously resets any pooled client before reusing the socket path, and GetOrCreateForType drops/reconnects a stale client when the backend identity changes. Added a real stopped-snapshot qemu -> qemu-microvm restore test that exercises the reused qemu.sock path.

File: lib/instances/create.go [hunk 2]
```diff
@@ -755,6 +748,17 @@ func (m *manager) startAndBootVM(
...
+	// qemu-microvm snapshots are tied to the binary that boots the VM. Refresh
+	// metadata on every cold start so host upgrades do not leave the instance's
+	// reported version pinned to its original creation time.
+	if stored.HypervisorType == hypervisor.TypeQEMUMicroVM {
+		detectedVersion, err := starter.GetVersion(m.paths)
+		if err != nil {
+			return fmt.Errorf("get QEMU version for qemu-microvm start: %w", err)
+		}
+		stored.HypervisorVersion = detectedVersion
+	}
```
[the remark this answers is not in the document]
Comment (rgarcia) [reply lib/instances/create.go:764]: Fixed in 33cffb15. Exact snapshot-version behavior is now declared by Capabilities.RequiresExactSnapshotVersion. Generic create, cold-start, and snapshot-target lifecycle code consults that capability instead of branching on TypeQEMUMicroVM.

File: lib/instances/snapshot_machine_type.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+			return machine, nil
+		}
+		resolved, err := qemu.ResolveMachineType("")
+		if err != nil {
+			return "", fmt.Errorf("%w: resolve QEMU machine type: %v", ErrInvalidRequest, err)
+		}
+		return resolved, nil
+	default:
+		return "", nil
+	}
```
[the remark this answers is not in the document]
Comment (rgarcia) [reply lib/instances/snapshot_machine_type.go]: Fixed in 8cc10e5d. The redundant persisted machine-type axis was removed, and stopped snapshot restore/fork targets now run the canonical qemu microvm validator before payload replacement or fork creation. TestStoppedSnapshotRejectsInvalidQEMUMicroVMTargetBeforeMutation verifies both no source mutation and no fork side effect.

File: lib/hypervisor/qemu/pool.go [hunk 4]
```diff
@@ -44,6 +57,19 @@ func GetOrCreate(socketPath string) (*QEMU, error) {
...
+// process reuses the same socket path.
+func resetClient(socketPath string) {
+	clientPool.Lock()
+	defer clientPool.Unlock()
+	if client, ok := clientPool.clients[socketPath]; ok {
+		delete(clientPool.clients, socketPath)
+		if client.client != nil {
+			_ = client.client.Close()
+		}
+	}
```
[the remark this answers is not in the document]
Comment (rgarcia) [reply lib/hypervisor/qemu/pool.go:80]: Fixed in the latest push. Pooled-client disconnects now happen outside the host-wide pool lock. Backend-mismatch reconnect releases the lock before closing and retries through normal pool arbitration.

[the remark this answers is not in the document]
Comment (rgarcia) [reply lib/hypervisor/qemu/pool.go:80]: Fixed in the latest push. QEMU error paths now call removeClient(q), which conditionally evicts by pointer generation. A stale client using a reused socket path cannot remove the replacement client. TestRemoveClientDoesNotEvictReplacement covers the ABA case, and the real qemu -> qemu-microvm stopped-snapshot switch test passes.

**Review by sjmiller609 — CHANGES_REQUESTED** [review]
**feedback**
* both Linux and Darwin CI failing
* lib/hypervisor/qemu/pool.go:34-49 — on type mismatch the pool trusts the caller and evicts the pooled client, so a caller holding stale metadata can close the live client. Since StartVM already calls resetClient before a new process claims the socket, a mismatch here can only mean a stale caller — return an error instead of replacing.

**architecture**

* lib/hypervisor/hypervisor.go:95 — make ValidateConfig part of VMStarter, this prevents changing outside of the interface unnecessarily
* model standard QEMU and microvm as profiles over the shared implementation:
    * TypeQEMU   → StandardProfile → shared QEMU
    * TypeQEMUMicroVM → MicroVMProfile  → shared QEMU
* lib/instances/hypervisor_version.go:11-46 — move host-version behavior into that profile/starter. Microvm create currently checks qemu --version up to three times
* lib/instances/vm_config_validation.go:9-53 — avoid duplicating base-disk and overlay counting from buildHypervisorConfig; share the calculation

**questions/test hardening**
* lib/hypervisor/qemu/qemu.go:54-67 — confirm why exact host-version compatibility applies only to microvm, not system-binary q35/virt snapshots.
* lib/instances/qemu_test.go:185-218 — fail rather than skip when QEMU/KVM is missing on the dedicated Linux KVM CI runner.

**Comment by rgarcia** [comment]
Addressed the requested changes in `4dd61a99`:

- QMP pool backend mismatches now return an error and never evict/close the live pooled client; explicit process startup remains responsible for `resetClient`.
- `ValidateConfig` is now a required `VMStarter` method, including tracing and all starter implementations.
- Standard QEMU and microvm are modeled as `StandardProfile` / `MicroVMProfile` policy over the shared process, QMP, snapshot, and restore implementation.
- Version selection moved behind `VMStarter.ResolveVersion`; QEMU create now executes `qemu --version` once and passes the resolved version through startup.
- Exact writer-version validation now applies to **both** system-binary QEMU profiles (`q35`/`virt` and `microvm`). Older standard-QEMU snapshots use the writer version already stored in generic instance metadata.
- Preflight and final VM configuration share `instanceDiskCount`, and final construction checks that the built disk list matches the plan.
- QEMU integration setup now opens `/dev/kvm` read/write and fails (rather than skips) when QEMU/KVM is unusable on Linux CI.
- Fixed the deterministic Linux/Darwin regressions in the existing image-retention and stopped-fork lock tests.

Local verification: targeted regression/profile/pool tests, QMP pool race tests, `go vet` on the touched packages, `go build ./...`, and compile-only `go test ./... -run '^$'`. The full Linux and Darwin jobs are queued on the new head.

**Comment by rgarcia** [comment]
Follow-up: Cursor Bugbot correctly found that the test-only `waitForQEMUReady` helper still selected the standard QEMU profile. Fixed in `aedb09ed` by passing the instance hypervisor type through every readiness call site; the Bugbot thread is resolved.

**Review by sjmiller609 — APPROVED** [review]
lgtm, after address issue with oapi gen and CI passing
